Random freeze at shutdown after successfully running pytest on Windows

It seems that we have a rare but unexpected failure in `_kill_process_tree_without_psutil` that can cause a worker to not being terminated correctly and possibly the cause  

```
022-02-21T18:01:14.5981714Z tests/test_worker_timeout.py::TestTimeoutExecutor::test_worker_timeout_shutdown_deadlock PASSED [100%]
2022-02-21T18:01:14.5983518Z 
2022-02-21T18:01:14.5984675Z ============================== warnings summary ===============================
2022-02-21T18:01:14.5986172Z tests/test_process_executor_spawn.py::TestsProcessPoolSpawnExecutor::test_worker_timeout
2022-02-21T18:01:14.5989879Z   D:\a\1\s\loky\process_executor.py:687: UserWarning: A worker stopped while some jobs were given to the executor. This can be caused by a too short worker timeout or by a memory leak.
2022-02-21T18:01:14.5991336Z     warnings.warn(
2022-02-21T18:01:14.5991918Z 
2022-02-21T18:01:14.5993128Z tests/test_reusable_executor.py::TestExecutorDeadLock::test_in_callback_submit_with_crash[id-args5-TerminatedWorkerError-SIGSEGV]
2022-02-21T18:01:14.5995035Z   D:\a\1\s\.tox\py38\lib\site-packages\_pytest\threadexception.py:73: PytestUnhandledThreadExceptionWarning: Exception in thread ExecutorManagerThread
2022-02-21T18:01:14.5997331Z   
2022-02-21T18:01:14.6000205Z   Traceback (most recent call last):
2022-02-21T18:01:14.6001647Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\threading.py", line 932, in _bootstrap_inner
2022-02-21T18:01:14.6003516Z       self.run()
2022-02-21T18:01:14.6005525Z     File "D:\a\1\s\loky\process_executor.py", line 550, in run
2022-02-21T18:01:14.6008559Z       self.terminate_broken(bpe)
2022-02-21T18:01:14.6009789Z     File "D:\a\1\s\loky\process_executor.py", line 735, in terminate_broken
2022-02-21T18:01:14.6010796Z       self.kill_workers(reason="broken executor")
2022-02-21T18:01:14.6011853Z     File "D:\a\1\s\loky\process_executor.py", line 765, in kill_workers
2022-02-21T18:01:14.6012855Z       kill_process_tree(p)
2022-02-21T18:01:14.6013779Z     File "D:\a\1\s\loky\backend\utils.py", line 19, in kill_process_tree
2022-02-21T18:01:14.6014847Z       _kill_process_tree_without_psutil(process)
2022-02-21T18:01:14.6015901Z     File "D:\a\1\s\loky\backend\utils.py", line 58, in _kill_process_tree_without_psutil
2022-02-21T18:01:14.6019063Z       subprocess.check_output(
2022-02-21T18:01:14.6020132Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 415, in check_output
2022-02-21T18:01:14.6021407Z       return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
2022-02-21T18:01:14.6022875Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 493, in run
2022-02-21T18:01:14.6023942Z       with Popen(*popenargs, **kwargs) as process:
2022-02-21T18:01:14.6025113Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 808, in __init__
2022-02-21T18:01:14.6028121Z       errread, errwrite) = self._get_handles(stdin, stdout, stderr)
2022-02-21T18:01:14.6029653Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 1199, in _get_handles
2022-02-21T18:01:14.6030862Z       errwrite = self._make_inheritable(errwrite)
2022-02-21T18:01:14.6031997Z     File "C:\hostedtoolcache\windows\Python\3.8.10\x64\lib\subprocess.py", line 1208, in _make_inheritable
2022-02-21T18:01:14.6033164Z       h = _winapi.DuplicateHandle(
2022-02-21T18:01:14.6034147Z   OSError: [WinError 6] The handle is invalid
2022-02-21T18:01:14.6034919Z   
2022-02-21T18:01:14.6037982Z     warnings.warn(pytest.PytestUnhandledThreadExceptionWarning(msg))
2022-02-21T18:01:14.6039599Z 
2022-02-21T18:01:14.6040397Z -- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
2022-02-21T18:01:14.6041642Z ----------------- generated xml file: D:\a\1\s\test-data.xml ------------------
2022-02-21T18:01:14.6042928Z =========== 226 passed, 28 skipped, 2 warnings in 518.81s (0:08:38) ===========
2022-02-21T18:01:14.6044058Z [INFO:MainProcess:MainThread] process shutting down
2022-02-21T18:01:14.6045243Z [DEBUG:MainProcess:MainThread] running all "atexit" finalizers with priority >= 0
2022-02-21T18:01:14.6048476Z [DEBUG:MainProcess:MainThread] Interpreter shutting down. Waking up executor_manager_thread []
2022-02-21T18:01:14.6049989Z [DEBUG:MainProcess:MainThread] telling queue thread to quit
2022-02-21T18:01:14.6051198Z [DEBUG:MainProcess:QueueFeederThread] feeder thread got sentinel -- exiting
2022-02-21T18:01:14.6052362Z [INFO:MainProcess:MainThread] calling join() for process LokyProcess-614
2022-02-21T18:51:19.4297519Z Terminate batch job (Y/N)? 
2022-02-21T18:51:19.4646241Z Entering debug mode. Use h or ? for help. 
2022-02-21T18:51:19.4647717Z 
2022-02-21T18:51:19.4671357Z At D:\a\_tasks\CmdLine_d9bafed4-0b18-4f58-968d-86655b4d2ce9\2.198.0\ps_modules\VstsTaskSdk\ToolFunctions.ps1:113 
2022-02-21T18:51:19.4672561Z char:13
2022-02-21T18:51:19.4672760Z 
2022-02-21T18:51:19.4673230Z +         if ($originalEncoding) {
2022-02-21T18:51:19.4673466Z 
2022-02-21T18:51:19.4674213Z +             ~~~~~~~~~~~~~~~~~
2022-02-21T18:51:20.3935136Z [DBG]: PS D:\a\1\s>> 
2022-02-21T18:51:20.3936071Z ^C
2022-02-21T18:51:20.4451106Z ##[error]The operation was canceled.
2022-02-21T18:51:20.4475233Z ##[section]Finishing: Test loky
```

Note that we catch `subprocess.CalledProcessError` but `OSError` was raised here.

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -11,7 +11,7 @@
 
 - Make `shutdown(kill_workers=True)` consistently use the SIGKILL
   signal on POSIX. Previously a mix of SIGKILL and SIGTERM was issued
-  and could deadlock the shutdown process (#348).
+  and could deadlock the shutdown process (#348 and #357).
 
 ### 3.0.0 - 2021-09-10
 

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -57,7 +57,7 @@ def _kill_process_tree_without_psutil(process):
             _windows_taskkill_process_tree(process.pid)
         else:
             _posix_recursive_kill(process.pid)
-    except Exception:
+    except Exception:  # pragma: no cover
         details = traceback.format_exc()
         warnings.warn(
             "Failed to kill subprocesses on this platform. Please install"
@@ -85,7 +85,7 @@ def _windows_taskkill_process_tree(pid):
         if e.returncode not in [128, 255]:
             # Let's raise to let the caller log the error details in a
             # warning and only kill the root process.
-            raise
+            raise  # pragma: no cover
 
 
 def _kill(pid):
@@ -101,7 +101,7 @@ def _kill(pid):
         # is already terminated, else, raise the error and let the top
         # level function raise a warning and retry to kill the process.
         if e.errno != errno.ESRCH:
-            raise
+            raise  # pragma: no cover
 
 
 def _posix_recursive_kill(pid):
@@ -115,7 +115,7 @@ def _posix_recursive_kill(pid):
         if e.returncode == 1:
             children_pids = ''
         else:
-            raise
+            raise  # pragma: no cover
 
     # Decode the result, split the cpid and remove the trailing line
     for cpid in children_pids.splitlines():

--- a/loky/backend/utils.py
+++ b/loky/backend/utils.py
@@ -5,6 +5,7 @@ import errno
 import signal
 import warnings
 import subprocess
+import traceback
 try:
     import psutil
 except ImportError:
@@ -51,39 +52,40 @@ def _kill_process_tree_with_psutil(process):
 
 def _kill_process_tree_without_psutil(process):
     """Terminate a process and its descendants."""
-    if sys.platform == "win32":
-        # On windows, the taskkill function with option `/T` terminate a given
-        # process pid and its children.
-        try:
-            subprocess.check_output(
-                ["taskkill", "/F", "/T", "/PID", str(process.pid)], stderr=None
-            )
-        except subprocess.CalledProcessError as e:
-            # In Windows, taskkill returns 1 for permission denied and 128, 255
-            # for no process found.
-            if e.returncode == 1:
-                # Try to terminate the process without its descendants if
-                # taskkill was denied permission. If this fails too, with an
-                # error different from process not found, let the top level
-                # function raise a warning and retry to kill the process.
-                process.terminate()
-            elif e.returncode in [128, 255]:
-                # Process not found: ignore.
-                pass
-            else:
-                raise
-    else:
-        try:
-            _recursive_kill_posix(process.pid)
-        except OSError:
-            warnings.warn(
-                "Failed to kill subprocesses on this platform. Please install"
-                "psutil: https://github.com/giampaolo/psutil"
-            )
-            # In case we cannot introspect the children, we fall back to only
-            # killing the main process.
-            process.kill()
+    try:
+        if sys.platform == "win32":
+            _windows_taskkill_process_tree(process.pid)
+        else:
+            _posix_recursive_kill(process.pid)
+    except Exception:
+        details = traceback.format_exc()
+        warnings.warn(
+            "Failed to kill subprocesses on this platform. Please install"
+            "psutil: https://github.com/giampaolo/psutil\n"
+            f"Details:\n{details}"
+        )
+        # In case we cannot introspect or kill the descendants, we fall back to
+        # only killing the main process.
+        #
+        # Note: on Windows, process.kill() is an alias for process.terminate()
+        # which in turns calls the Win32 API function TerminateProcess().
+        process.kill()
     process.join()
+
+
+def _windows_taskkill_process_tree(pid):
+    # On windows, the taskkill function with option `/T` terminate a given
+    # process pid and its children.
+    try:
+        subprocess.check_output(
+            ["taskkill", "/F", "/T", "/PID", str(pid)], stderr=None
+        )
+    except subprocess.CalledProcessError as e:
+        # In Windows, taskkill returns 128, 255 for no process found.
+        if e.returncode not in [128, 255]:
+            # Let's raise to let the caller log the error details in a
+            # warning and only kill the root process.
+            raise
 
 
 def _kill(pid):
@@ -102,7 +104,7 @@ def _kill(pid):
             raise
 
 
-def _recursive_kill_posix(pid):
+def _posix_recursive_kill(pid):
     """Recursively kill the descendants of a process before killing it."""
     try:
         children_pids = subprocess.check_output(
@@ -118,7 +120,7 @@ def _recursive_kill_posix(pid):
     # Decode the result, split the cpid and remove the trailing line
     for cpid in children_pids.splitlines():
         cpid = int(cpid)
-        _recursive_kill_posix(cpid)
+        _posix_recursive_kill(cpid)
 
     _kill(pid)
 

